http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/AwaitCloseChannelPoolMap.java

```diff
@@ -0,0 +1,250 @@
+import java.util.concurrent.atomic.AtomicReference;
+import javax.net.ssl.SSLException;
+import javax.net.ssl.TrustManagerFactory;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.annotations.SdkTestInternalApi;
+import software.amazon.awssdk.http.Protocol;
+import software.amazon.awssdk.http.nio.netty.SdkEventLoopGroup;
+import software.amazon.awssdk.http.nio.netty.internal.http2.HttpOrHttp2ChannelPool;
+import software.amazon.awssdk.utils.Logger;
+
+/**
+ * Implementation of {@link SdkChannelPoolMap} that awaits channel pools to be closed upon closing.
+ */
+@SdkInternalApi
+public final class AwaitCloseChannelPoolMap extends SdkChannelPoolMap<URI,
+    AwaitCloseChannelPoolMap.SimpleChannelPoolAwareChannelPool> {
+
+    private static final Logger log = Logger.loggerFor(AwaitCloseChannelPoolMap.class);
+
+    private final SdkChannelOptions sdkChannelOptions;
+    private final SdkEventLoopGroup sdkEventLoopGroup;
+    private final NettyConfiguration configuration;
+    private final Protocol protocol;
+    private final long maxStreams;
+    private final SslProvider sslProvider;
+
+    private AwaitCloseChannelPoolMap(Builder builder) {
+        this.sdkChannelOptions = builder.sdkChannelOptions;
+        this.sdkEventLoopGroup = builder.sdkEventLoopGroup;
+        this.configuration = builder.configuration;
+        this.protocol = builder.protocol;
+        this.maxStreams = builder.maxStreams;
+        this.sslProvider = builder.sslProvider;
+    }
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    @Override
+    protected SimpleChannelPoolAwareChannelPool newPool(URI key) {
+        SslContext sslContext = sslContext(key.getScheme());
+        Bootstrap bootstrap =
+            new Bootstrap()
+                .group(sdkEventLoopGroup.eventLoopGroup())
+                .channelFactory(sdkEventLoopGroup.channelFactory())
+                .option(ChannelOption.CONNECT_TIMEOUT_MILLIS, configuration.connectTimeoutMillis())
+                // TODO run some performance tests with and without this.
+                .remoteAddress(key.getHost(), key.getPort());
+        sdkChannelOptions.channelOptions().forEach(bootstrap::option);
+
+        AtomicReference<ChannelPool> channelPoolRef = new AtomicReference<>();
+        ChannelPipelineInitializer handler =
+            new ChannelPipelineInitializer(protocol, sslContext, maxStreams, channelPoolRef, configuration, key);
+
+        BetterSimpleChannelPool simpleChannelPool = new BetterSimpleChannelPool(bootstrap, handler);
+
+        channelPoolRef.set(wrapSimpleChannelPool(bootstrap, simpleChannelPool));
+        return new SimpleChannelPoolAwareChannelPool(simpleChannelPool, channelPoolRef.get());
+    }
+
+    @Override
+    public void close() {
+        log.trace(() -> "Closing channel pools");
+        // If there is a new pool being added while we are iterating the pools, exception could be thrown.
```

Comment:
If a new pool is acquired, is there a possibility of it creating a channel that won't get closed?

Reply:
The new pool would be closed since we are calling `super.close()`, but there is a possibility that the same race condition could occur between the newlyAcquiredChannelPool.close and eventloop.shutdown

Reply:
Is it possible a new pool is created while `super.close()` is running though? It doesn't look like access to the underlying map is synchronized.

Reply:
Yeah, it's possible. There's no way to synchronize retrieval operations with `ConcurrentHashMap` and there's going to be performance degradation if we switch to synchronized map. I'm not too concerned about this rare edge case.

Reply:
Okay. Can we at least note it? Maybe we can find a way to fill in this gap down the line that won't affect performance.

Reply:
Yeah, added notes.